src/commands.rs

```diff
@@ -593,7 +593,34 @@ pub fn run_command(cmd: Command) -> Result<i32> {
                 }
                 ServerStartup::TimedOut => bail!("Timed out waiting for server startup"),
                 ServerStartup::AddrInUse => bail!("Server startup failed: Address in use"),
-                ServerStartup::Err { reason } => bail!("Server startup failed: {}", reason),
+                ServerStartup::Err { reason } => {
+                    let os_10013 = reason.as_str().contains("os error 10013");
```

Comment:
A `SeverStartup::AddrInUse` only happens when the port is bound. The problem with 10013 is that the port is in an excluded range, not that it is or isn't bound. The applications are using [`CreatePersistentTcpPortReservation`](https://docs.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-createpersistenttcpportreservation) to reserve an entire range of TCP ports and only binding to them when they choose. The reservation makes sure nobody else can take your port range, even if the ports were configured for reusable binding. Instead of using the command line I could just call into that function from Rust using FFI. I know winapi is already a dependency, and it includes the `iphlpapi.h` header defines as "iphlpapi"

So with 
```

[target.'cfg(windows)'.dev-dependencies.winapi]
version = "0.3"
features = [
    "iphlpapi"
]
```

I could write a test that uses the appropriate functions and just marks itself as a pass if it gets a `ERROR_ACCESS_DENIED` result (with maybe a warning message about some tests being unable to run without admin permissions).

Reply:
> A SeverStartup::AddrInUse only happens when the port is bound.

`ServerStartup` is defined by us in `sccache` itself, so we control which OS errors get mapped to it.
[Currently, `io::ErrorKind::AddrInUse` the only kind of error mapped to `ServerStartup::AddrInUse`](https://github.com/mozilla/sccache/blob/6488b3c7fd7145940d16e52fe1022a9249dad696/src/server.rs#L426), but we can modify this code.

> The applications are using [CreatePersistentTcpPortReservation](https://docs.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-createpersistenttcpportreservation) to reserve an entire range of TCP ports and only binding to them when they choose. The reservation makes sure nobody else can take your port range, even if the ports were configured for reusable binding. Instead of using the command line I could just call into that function from Rust using FFI. I know winapi is already a dependency, and it includes the iphlpapi.h header defines as "iphlpapi"

I don't quite follow you here: I'm not sure if you're referring to the tests or the "is port already in-use/allocated/reserved/not-accessible" behaviour at runtime.
I _don't_ think that we should be calling `[CreatePersistentTcpPortReservation](https://docs.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-createpersistenttcpportreservation), because we only need the one port, not multiple - and we still have to handle the same `10013` error anyways.

> I could write a test that uses the appropriate functions and just marks itself as a pass if it gets a ERROR_ACCESS_DENIED result (with maybe a warning message about some tests being unable to run without admin permissions).

This test would be flaky depending on whether it's running on a system with WSL/Hyper-V, or if it has admin permissions. I don't think it's worth it personally.

Reply:
There is no easy way to implement the errnum check in lower code as far as I can see. To keep the performance benefit I moved the code into Windows specific implementation of a function. I also added a guard statement so that if the String comparison breaks or changes form (which should be impossible unless we start generating non OS error messages) we get a Debug message telling the dev to check the statements for consistency. It shouldn't break though since `Error::os_last_error().into()` generates a `String` the same way that `Error::from_raw_os_error(u16).to_string()` does.

Reply:
So I looked into what `ErrorKind` is issued for the 10013 errnum, and it's an `ErrorKind::Uncategorized`. As a result I think that taking the `ErrorKind` approach is really not that useful, since we have to rely on unstable Rust.

Reply:
> There is no easy way to implement the errnum check in lower code as far as I can see.

I don't understand, why can't we edit the code [where we map from `io::Error` to `ServerStartup`](https://github.com/mozilla/sccache/blob/6488b3c7fd7145940d16e52fe1022a9249dad696/src/server.rs#L426)?
I'm guessing it'd look something like this:

```
match e.downcast_ref::<io::Error>() {
    Some(io_err) if io::ErrorKind::AddrInUse == io_err.kind() => {
        notify_server_startup(&notify, ServerStartup::AddrInUse)?;
    }
    Some(io_err) if cfg!(windows) && Some(10013) == io_err.raw_os_error() => {
        // 10013 is the "WSAEACCES" error, which can occur if the requested port
        // has been allocated for other purposes, such as winNAT or Hyper-V.
        notify_server_startup(&notify, ServerStartup::AddrInUse)?;
    }
    _ => {
        let reason = e.to_string();
        notify_server_startup(&notify, ServerStartup::Err { reason })?;
    }
};
```

Does that make sense?